src/Driver/PDO/Statement.php

```diff
@@ -117,12 +133,109 @@ public function execute($params = null): ResultInterface
             );
         }
 
+        $resourceOffsets = $this->getResourceOffsets();
         try {
             $this->stmt->execute($params);
         } catch (PDOException $exception) {
             throw Exception::new($exception);
+        } finally {
+            if ($resourceOffsets !== null) {
+                $this->restoreResourceOffsets($resourceOffsets);
+            }
         }
 
         return new Result($this->stmt);
     }
+
+    /**
+     * Track a binary parameter reference at binding time. These
+     * are cached for later analysis by the getResourceOffsets.
+     *
+     * @param mixed $resource
+     */
+    private function trackParamResource($resource): void
+    {
+        if (! is_resource($resource)) {
+            return;
+        }
+
+        $resources   = &$this->paramResources;
+        $resources ??= [];
+        $resources[] = $resource;
+    }
+
+    /**
+     * Determine the offset that any resource parameters needs to be
+     * restored to after the statement is executed. Call immediately
+     * before execute (not during bindValue) to get the most accurate offset.
+     *
+     * @return int[]|null Return offsets to restore if needed.
+     */
+    private function getResourceOffsets()
+    {
+        $resources = &$this->paramResources;
+        if ($resources === null) {
+            return null;
+        }
+
+        $count           = count($resources);
+        $resourceOffsets = null;
+        $deref           = [];
+        for ($i = 0; $i < $count; ++$i) {
+            $resource = $resources[$i];
+            $position = false;
+            try {
+                if (stream_get_meta_data($resource)['seekable']) {
+                    $position = ftell($resource);
+                }
+            } catch (Throwable $e) {
+                // Swallow
+            } finally {
+                if ($position === false) {
+                    if ($resourceOffsets !== null) {
+                        $resourceOffsets[] = -1;
+                    }
+                } else {
+                    $resourceOffsets ??= $i !== 0 ? array_fill(0, $i, -1) : [];
+                    $resourceOffsets[] = $position;
+                }
+            }
+        }
+
+        if ($resourceOffsets === null) {
+            $resources = null;
+        }
+
+        return $resourceOffsets;
+    }
+
+    /**
+     * Restore resource offsets moved by PDOStatement->execute
+     *
+     * @param int[]|null $resourceOffsets The offsets returned by getResourceOffsets.
+     */
+    private function restoreResourceOffsets($resourceOffsets): void
```

Comment:
```suggestion
    private function restoreResourceOffsets(?array $resourceOffsets): void
```

Reply:
The @param int[]|null doc comment already provides a stronger type than ?array. If you want both, fine, but why isn't this redundant or even contradictory? I was concerned about providing two types that are not equivalent. I'm still learning the php typing system.

Reply:
> If you want both,

I do. 🙂 

> why isn't this redundant or even contradictory?

Native types are checked at runtime, so we know when they start to smell. `?array` is as precise as we can get with native types, so we should take it.

PHPDoc types are read by static analysis tools (and humans of course) and they allow us to be way more precise than the native type system does. Common static analyzers that that into account and allow the PHPDoc type to be narrower than the native one.